### ISSUE TITLE
Stats reports 1 cache miss although none is found in debug build dir

### ISSUE THREAD

**Issue by Delgan** [ccache/ccache#1088]
Hi.

I'm not sure if this is truly a bug report of more a question, sorry. 

### How to reproduce ###

Obviously it won't necessary work in your case but here is what I'm doing:

1. Configure ccache with debug enabled
2. Run compilation once
3. Run compilation again (no change)
4. Call `ccache -vs`

### Actual behavior ###

Ccache stats:

```log
 Summary:
  Cache directory:          /root/.ccache
  Primary config:           /root/.ccache/ccache.conf
  Secondary config:         /usr/local/etc/ccache.conf
  Stats updated:            Fri Jun  3 20:49:48 2022
  Hits:                      878 /  879 (99.89 %)
    Direct:                  878 /  879 (99.89 %)
    Preprocessed:              0 /    1 (0.00 %)
  Misses:                      1
    Direct:                    1
    Preprocessed:              1
  Uncacheable:               242
Primary storage:
  Hits:                        0 / 1758 (0.00 %)
  Misses:                   1758
  Cache size (GB):          0.00 / 5.00 (0.00 %)
  Files:                       2
Secondary storage:
  Hits:                     1756 / 1758 (99.89 %)
  Misses:                      2
Uncacheable:
  Called for linking:        119
  Called for preprocessing:  120
  No input file:               3
```
There is 1 cache miss. I downloaded the debug dir generated by Ccache, then I searched each file for `"cache_miss"` but found none. Also searched for [`"No \S+ in http*"`](https://github.com/ccache/ccache/blob/0d60322086c0030bb1af72891d48e83be5c295a2/src/storage/Storage.cpp#L495-L498) with no results.
I tried to rebuild several times but could never reach 100%.

Every compilation unit output is either:

```log
[2022-06-03T19:33:52.853181 2697 ] Result: direct_cache_hit
[2022-06-03T19:33:52.853186 2697 ] Result: primary_storage_miss
[2022-06-03T19:33:52.853189 2697 ] Result: secondary_storage_hit
```

Or for link step:

```log
[2022-06-03T19:33:43.937205 2314 ] Result: called_for_link
```

How can I investigate which file is failing?

### Expected behavior ###

100% cache hit **OR** cache miss data in the debug directory.

<s>Also I don't know if it's related, but despite 3 `"No input file"` reported, I find not any [`"No input file found"`](https://github.com/ccache/ccache/blob/0d60322086c0030bb1af72891d48e83be5c295a2/src/argprocessing.cpp#L1169) or [`"No profile data file found"`](https://github.com/ccache/ccache/blob/cbedccc9fa30f8958dc3e7fbdf900a952f63d5ad/src/ccache.cpp#L1738) in logs. Some others messages from the `LOG_RAW()` macro are fine, though.</s>
Well, at there is no input file, there is no output file neither to store the logs. :)
I found the messages after adding the `log_file` config variable.

### Environment ###

CCache 4.6.0
Ubuntu 18.04
CMake 3.19.5
GCC 8.4.0
Gold linker 1.15
Ninja 1.8.2

Configuration:

```ini
hash_dir = false
compiler_check = content
sloppiness = include_file_ctime,include_file_mtime
secondary_storage = http://ccache-remote-storage.storage.googleapis.com|bearer-token=*******|share-hits=false
debug = true
debug_dir = ccache_debug
log_file = ccache_debug.log
```

Ccache built with:

```
-DCMAKE_BUILD_TYPE=Release \
-DREDIS_STORAGE_BACKEND=OFF \
-DENABLE_DOCUMENTATION=OFF \
-DENABLE_TESTING=OFF \
```

**Comment by Delgan** [comment]
@jrosdahl No unfortunately. 3 first files analyzed resulted in "`No input file"` in the global `.log` file. All subsequents files where `"Result: direct_cache_hit"` or `"Result: called_for_link"`.

**Comment by jrosdahl** [comment]
There are cases were you won't see a cache_miss log in a debug log file produced in `debug_dir`, for example if the build first compiles an object file resulting in a cache miss and then compiles the same file again since the debug log file  will be overwritten and say that it was a cache hit. However, the ordinary `log_file` captures everything and it seems quite unlikely that there wouldn't be any log line for a cache miss. Are you sure you captured the full build in the log?

If you want more help I think you need to come up with an example that I can run to reproduce the problem.

**Comment by jrosdahl** [comment]
> I think I finally managed to provide a minimal reproducible example!

Great! However, something is missing from your reproduction instructions:

```console
% cat CMakeLists.txt 
include(FetchContent)

cmake_minimum_required(VERSION 3.17 FATAL_ERROR)
project(DebugCCache VERSION 1.0.0)

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -fprofile-arcs -ftest-coverage")

set(valijson_BUILD_TESTS OFF CACHE BOOL "" FORCE)
set(valijson_EXCLUDE_BOOST ON CACHE BOOL "" FORCE)
FetchContent_Declare(
  valijson
  GIT_REPOSITORY https://github.com/tristanpenman/valijson.git
  GIT_TAG        4897d102bd8561ae2771f0bb879a2ded036d0761
)
FetchContent_MakeAvailable(valijson)
% mkdir build
% cd build
% cmake -DCMAKE_BUILD_TYPE=Debug ..                                               
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is GNU 9.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/joel/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/joel/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test COMPILER_SUPPORTS_CXX11
-- Performing Test COMPILER_SUPPORTS_CXX11 - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /home/joel/tmp/ccache-issue-1088/build
% ninja -j1
ninja: no work to do.
```

> The ccache.log file (no "cache_miss" reference):

The log file you included only contains two invocations (both with `Result: no_input_file`), but the statistics say that there were 10 calls (3 cacheable and 5+2 uncacheable), so you seem to have missed lots of invocations.

I notice that you have `log_file = ccache.log` in your config file. That will make ccache print to `ccache.log` in the current working directory, which depends on where the compiler is invoked, so I suspect that you get multiple different ccache.log files. Try setting `log_file = /tmp/ccache.log` or so instead.

**Comment by Delgan** [comment]
So that was it... I didn't realize that several log files could be generated in different places.
I searched once inside the `build/_deps` folder generated by `FetchContent()` and found nothing, so I wrongly assumed that all logs were in the same place.
Actually, while building dependency CMake uses a temporary directory which is then deleted, and I suppose the CCache logs for the dependency were removed at the same time.
The `ninja -1` in the example was not even needed as the cache miss is generated by CMake `FetchContent()`.

Now that I used an absolute paths for the logs, I finally spotted the unique `cache_miss` I was hunting for.
I just have to dig into this now, to finally reach the 100% cache hit. :)

Thanks a lot for your guidance and sorry for the noise. :+1:
